src/ipc/capnp/common-types.h

```diff
@@ -0,0 +1,281 @@
+
+template <typename T>
+struct Unserializable
+{
+private:
+    template <typename C>
+    static std::true_type test(decltype(std::declval<C>().Unserialize(std::declval<C&>()))*);
+    template <typename>
+    static std::false_type test(...);
+
+public:
+    static constexpr bool value = decltype(test<T>(nullptr))::value;
+};
+
+template <typename T>
+struct Serializable
+{
+private:
+    template <typename C>
+    static std::true_type test(decltype(std::declval<C>().Serialize(std::declval<C&>()))*);
+    template <typename>
+    static std::false_type test(...);
+
+public:
+    static constexpr bool value = decltype(test<T>(nullptr))::value;
+};
+} // namespace capnp
+} // namespace ipc
+
+namespace mp {
+//!@{
+//! Functions to serialize / deserialize bitcoin objects that don't
+//! already provide their own serialization.
+void CustomBuildMessage(InvokeContext& invoke_context,
+                        const UniValue& univalue,
+                        ipc::capnp::messages::UniValue::Builder&& builder);
+void CustomReadMessage(InvokeContext& invoke_context,
+                       const ipc::capnp::messages::UniValue::Reader& reader,
+                       UniValue& univalue);
+//!@}
+
+template <typename LocalType, typename Reader, typename ReadDest>
+decltype(auto) CustomReadField(
+    TypeList<LocalType>,
+    Priority<2>,
+    InvokeContext& invoke_context,
+    Reader&& reader,
+    ReadDest&& read_dest,
+    decltype(CustomReadMessage(invoke_context, reader.get(), std::declval<LocalType&>()))* enable = nullptr)
+{
+    return read_dest.update([&](auto& value) { CustomReadMessage(invoke_context, reader.get(), value); });
+}
+
+template <typename Accessor, typename LocalType, typename ServerContext, typename Fn, typename... Args>
+auto CustomPassField(TypeList<LocalType>, ServerContext& server_context, Fn&& fn, Args&&... args)
+    -> decltype(CustomPassMessage(server_context,
+                                  Accessor::get(server_context.call_context.getParams()),
+                                  Accessor::init(server_context.call_context.getResults()),
+                                  nullptr))
+{
+    CustomPassMessage(server_context, Accessor::get(server_context.call_context.getParams()),
+                      Accessor::init(server_context.call_context.getResults()),
+                      [&](LocalType param) { fn.invoke(server_context, std::forward<Args>(args)..., param); });
+}
+
+template <typename LocalType, typename Input, typename ReadDest>
+decltype(auto) CustomReadField(
+    TypeList<LocalType>,
+    Priority<1>,
+    InvokeContext& invoke_context,
+    Input&& input,
+    ReadDest&& read_dest,
+    typename std::enable_if<ipc::capnp::Deserializable<LocalType>::value>::type* enable = nullptr)
+{
+    assert(input.has());
+    auto data = input.get();
+    // Note: stream copy here is unnecessary, and can be avoided in the future
+    // when `VectorReader` from #12254 is added.
```

Comment:
eae99d609c7a08fde35e723967d078904ad1cd05  That seems a different Bitcoin Core PR... however there have been a few VectorReader merges recently.